core/primitives/src/epoch_manager.rs

```diff
@@ -104,32 +109,63 @@ pub struct AllEpochConfig {
 impl AllEpochConfig {
     pub fn new(
         use_production_config: bool,
+        genesis_protocol_version: ProtocolVersion,
         genesis_epoch_config: EpochConfig,
         chain_id: &str,
     ) -> Self {
-        Self {
+        Self::new_with_test_overrides(
             use_production_config,
+            genesis_protocol_version,
             genesis_epoch_config,
-            chain_id: chain_id.to_string(),
-            test_overrides: AllEpochConfigTestOverrides::default(),
-        }
+            chain_id,
+            Some(AllEpochConfigTestOverrides::default()),
+        )
     }
 
     pub fn new_with_test_overrides(
         use_production_config: bool,
+        genesis_protocol_version: ProtocolVersion,
         genesis_epoch_config: EpochConfig,
         chain_id: &str,
         test_overrides: Option<AllEpochConfigTestOverrides>,
     ) -> Self {
-        Self {
+        // Use the config store only for production configs and outside of tests.
+        let config_store = if use_production_config && test_overrides.is_none() {
+            EpochConfigStore::for_chain_id(chain_id)
+        } else {
+            None
+        };
+        let all_epoch_config = Self {
+            config_store: config_store.clone(),
             use_production_config,
             genesis_epoch_config,
             chain_id: chain_id.to_string(),
             test_overrides: test_overrides.unwrap_or_default(),
+        };
+        // Sanity check: Validate that the stored genesis config equals to the config generated for the genesis protocol version.
+        // Note that we cannot do this in unittests because we do not have direct access to the genesis config for mainnet/testnet.
+        // Thus, by making sure that the generated and store configs match for the genesis, we complement the unittests, which
+        // check that the generated and stored configs match for the versions after the genesis.
+        if config_store.is_some() {
+            debug_assert_eq!(
+                config_store.as_ref().unwrap().get_config(genesis_protocol_version).as_ref(),
+                &all_epoch_config.generate_epoch_config(genesis_protocol_version),
+                "Provided genesis EpochConfig for protocol version {} does not match the stored config", genesis_protocol_version
+            );
```

Comment:
nit: giving the first config the number `GENESIS_VERSION.json` instead of `0.json` and exposing it in `EpochConfigStore::genesis_protocol_version` could be a bit leaner and more correct - because technically for small protocol version number (1, 2...) epoch config doesn't exist and trying to get it is a logical error.

Reply:
done. set genesis version from 0 to 29. Also tested it in mainnet and testnet to make sure the genesis epoch infos and other infos match.